lib/Translation/ExportVerilog/ExportVerilog.cpp

```diff
@@ -2890,52 +2892,79 @@ struct RootEmitterBase {
+        llvm::sys::path::append(outputPath, directory.getValue());
 
-    // Otherwise implicitly assign the operation to a dedicated output file or
-    // mark it as a replicated operation.
-    auto maybeSeparate = [&](Operation *op, Twine fileName) {
-      if (separateModules) {
-        SmallString<32> fileNameStr;
-        fileName.toVector(fileNameStr);
-        auto fileNameAttr = Identifier::get(fileNameStr, op->getContext());
-        auto &file = files[fileNameAttr];
-        file.ops.push_back(info);
-      } else {
-        rootFile.ops.push_back(info);
+      if (auto name = attr.name()) {
+        llvm::sys::path::append(outputPath, name.getValue());
+        hasFileName = true;
       }
+
+      emitReplicatedOps = !attr.exclude_replicated_ops().getValue();
+      addToFilelist = !attr.exclude_from_filelist().getValue();
+    }
+
+    auto separateFile = [&](Operation *op, Twine fileName = "") {
+      if (!hasFileName)
+        llvm::sys::path::append(outputPath, fileName);
+
+      auto &file = files[Identifier::get(outputPath, op->getContext())];
+      file.ops.push_back(info);
+      file.emitReplicatedOps = emitReplicatedOps;
+      file.addToFilelist = addToFilelist;
     };
 
+    // Separate the operation into dedicated output file, or emit into the root
+    // file, or replicate in all output files.
     TypeSwitch<Operation *>(&op)
         .Case<HWModuleOp>([&](auto &mod) {
           // Emit into a separate file named after the module.
-          maybeSeparate(mod, mod.getName() + ".sv");
+          if (attr || separateModules)
+            separateFile(mod, mod.getName() + ".sv");
+          else
+            rootFile.ops.push_back(info);
         })
         .Case<InterfaceOp>([&](auto &intf) {
           // Emit into a separate file named after the interface.
-          maybeSeparate(intf, intf.sym_name() + ".sv");
+          if (attr || separateModules)
+            separateFile(intf, intf.sym_name() + ".sv");
+          else
+            rootFile.ops.push_back(info);
         })
         .Case<VerbatimOp, IfDefProceduralOp, TypeScopeOp, HWModuleExternOp>(
             [&](auto &) {
-              // Replicate in each outputfile.
-              replicatedOps.push_back(&op);
+              // Emit into a separate file using the specified file name or
+              // replicate the operation in each outputfile.
+              if (attr) {
+                if (!hasFileName) {
+                  op.emitError("file name unspecified");
+                  encounteredError = true;
+                } else
+                  separateFile(&op);
+              } else
+                replicatedOps.push_back(&op);
             })
         .Case<HWGeneratorSchemaOp>([&](auto &) {
-          // Empty.
+          if (attr) {
+            if (!hasFileName) {
+              op.emitError("file name unspecified");
+              encounteredError = true;
+            } else
+              separateFile(&op);
+          }
```

Comment:
I don't think we need a filename/separate file op for the generator schemas, right @darthscsi?

Reply:
Schema should never appear in the output.